### PR TITLE
REF: Hardcode Keras params

### PR DESCRIPTION
**Pull request by adriangb**
This implements hardcoding of common Keras parameters in `BaseWrapper.__init__` as discussed in #37.

The parameters chosen come from `Model.compile`, `Model.predict` and `Model.fit`. Parameters were not hardcoded if:
- They don't make sense for Scikit-Learn/Numpy arrays (for example, `steps_per_epoch`).
- Should be fed to `fit` directly because they depend on the data (ex: `sample_weights`, `class_weights`). I think `validation_data` also falls into this category, but I am not sure if this makes much sense to implement since AFAIK this is not used anywhere else in Scikit-Learn. I am going to leave this for another PR where I'll also deprectate `fit`'s `**kwargs`.

### CODE REVIEW

File: tests/test_wrappers.py [hunk 1]
```diff
@@ -1224,24 +1223,6 @@ def testpostprocess_y(self):
...
-        class ClassBuildFnClf(wrappers.KerasClassifier):
-            def __init__(self, input_param):
-                # does not set input_param
-                super().__init__()
-
-            def _keras_build_fn(self, hidden_dim):
-                return build_fn_clf(hidden_dim)
-
-        with pytest.raises(RuntimeError):
-            ClassBuildFnClf(input_param=10)
```
Comment (adriangb) [inline tests/test_wrappers.py:1242]: Removing this check and test. It is really a redundant feature that no other estimators have. It was much more helpful previously when the interface was less defined.

File: tests/test_wrappers.py [hunk 2]
```diff
@@ -417,9 +417,12 @@ class ClassWithCallback(wrappers.KerasClassifier):
...
     """
 
-    def __init__(self, **sk_params):
-        self.callbacks = [SentinalCallback()]
-        super().__init__(**sk_params)
+    def __init__(self, callbacks=None, hidden_dim=(100,)):
+        if callbacks is None:
+            self.callbacks = [SentinalCallback()]
+        else:
+            self.callbacks = callbacks
```
Comment (adriangb) [inline tests/test_wrappers.py:424]: I think this brings up an interesting question that is not a blocker for this PR but will be interesting for param routing: a quite common pattern in scikit-learn estimators is to parse/modify an `__init__` param in `fit`, ex:
```python
def fit(X, y, ...):
    self.random_state_ = np.random.RandomState(seed=self.random_state)
```

Where downstream the estimator will only use the `self.random_state_` param and not `self.random_state`.

I think it would be nice to allow users to do something like:

```python3
class MyClassifier(KerasClassifier):

    def hook_params(self):   # in base class this would just be a pass
         if self.batch_size is None:
              self.batch_size_ = 32
         else:
              self.batch_size_ = self.batch_size
```

But the trick will then be routing that parameter so that `callbacks_` always overrides `callbacks`. Maybe if two parameters match except for `_`'s at the end, the one with the most `_` takes precedence?  So that:

```python3
def build_fn(batch_size=32):
    print(f"batch_size={batch_size}")
    ...

class MyClassifier(KerasClassifier):

    def hook_params(self):   # in base class this would just be a pass
         if self.batch_size is None:
              self.batch_size_ = 32
         else:
              self.batch_size_ = self.batch_size
         self.batch_size__ = 64

estimator = MyClassifier()
estimator.fit(X, y)  # always prints `batch_size=64`
```

File: scikeras/wrappers.py [hunk 3]
```diff
@@ -676,11 +638,11 @@ def _filter_params(self, fn, params_to_check=None):
...
-        parameters = super()._get_param_names()
-        # add kwargs/sk_params if the user gave those as input
-        if self._sk_params:
-            return sorted(parameters + self._sk_params)
-        return parameters
+        return (
+            k
+            for k in self.__dict__
+            if not k.endswith("_") and not k.startswith("_")
+        )
```
Comment (adriangb) [inline scikeras/wrappers.py:639]: This is borrowed from SciKeras. For where we are headed, I think this makes more sense than storing `_sk_params`. It'll be up to users to make sure they don't set any params without `_` that they don't want to show up in `get_params`.

File: scikeras/wrappers.py [hunk 4]
```diff
@@ -51,94 +41,67 @@ class BaseWrapper(BaseEstimator):
...
+        self.batch_size = batch_size
+        self.verbose = verbose
+        self.callbacks = callbacks
+        self.epochs = epochs
+        self.validation_split = validation_split
+        self.shuffle = shuffle
+        self.initial_epoch = initial_epoch
+
+        # Unpack kwargs
+        vars(self).update(**kwargs)
```
Comment (adriangb) [inline scikeras/wrappers.py:104]: Also borrowed from scikeras. So much cleaner with the new `_get_param_names` implementation.

**Comment by adriangb** [comment]
tagging @stsievert in case you want to take a look

File: tests/test_wrappers.py (hunk 2 above)
Comment (stsievert) [reply tests/test_wrappers.py:424]: Fyi, sklearn.utils has a "check_random_state" function. I don't recall estimators ever setting a "random_state_" attribute.

Comment (adriangb) [reply tests/test_wrappers.py:424]: Yeah, I was just looking for a clear example, I didn't check that it's a real one.

Comment (adriangb) [reply tests/test_wrappers.py:424]: I updated it to a real-world example from Scikit-Learn.

Comment (stsievert) [reply tests/test_wrappers.py:424]: Thanks for the update; I also do that in my estimators.

Comment (adriangb) [reply tests/test_wrappers.py:424]: Again, not for this PR, but certainly a required feature for parameter routing. I'm going to copy this discussion to #50 .

File: scikeras/wrappers.py (hunk 3 above)
Comment (adriangb) [reply scikeras/wrappers.py:631]: I added `k.startswith("_")` to accomodate parameters like `self._random_state`. So the rule will be that all model paramers (i.e. the `__init__` ones) cannot start/end with `_`, at least for now. We'll have to see how that plays in with my overwriting proposal in https://github.com/adriangb/scikeras/pull/47#discussion_r469709646

**Comment by adriangb** [comment]
@stsievert do you think we are good to go here or is there anything else you see that should be changed?

File: scikeras/wrappers.py [hunk 5]
```diff
@@ -51,102 +41,67 @@ class BaseWrapper(BaseEstimator):
...
+        random_state=None,
+        optimizer="rmsprop",
+        loss=None,
+        metrics=None,
+        run_eagerly=None,
+        batch_size=None,
+        verbose=1,
+        steps=None,
+        callbacks=None,
+        epochs=1,
```
Comment (stsievert) [inline scikeras/wrappers.py:74]: Maybe rename to `max_epochs` for the `fit` implementation. I think `partial_fit` should always have `epochs=1`.

File: scikeras/wrappers.py [hunk 6]
```diff
@@ -51,102 +41,67 @@ class BaseWrapper(BaseEstimator):
...
+        validation_split=0.0,
+        shuffle=True,
+        initial_epoch=0,
```
Comment (stsievert) [inline scikeras/wrappers.py:77]: This only makes sense if a pretrained model is passed to `build_fn`. I'd be okay removing support for this, especially if `max_epochs` is supported.

File: scikeras/wrappers.py [hunk 7]
```diff
@@ -51,102 +41,67 @@ class BaseWrapper(BaseEstimator):
...
-    def __init__(self, build_fn=None, **sk_params):
+    def __init__(
+        self,
+        build_fn=None,
+        *,
+        random_state=None,
+        optimizer="rmsprop",
+        loss=None,
+        metrics=None,
+        run_eagerly=None,
```
Comment (stsievert) [inline scikeras/wrappers.py:69]: Does SciKeras have tests for `run_eagerly in [True, False]`? It seems `run_eagerly` will affect the serialization.

File: scikeras/wrappers.py [hunk 8]
```diff
@@ -51,102 +41,67 @@ class BaseWrapper(BaseEstimator):
...
     }
 
-    def __init__(self, build_fn=None, **sk_params):
+    def __init__(
+        self,
+        build_fn=None,
+        *,
+        random_state=None,
+        optimizer="rmsprop",
+        loss=None,
```
Comment (stsievert) [inline scikeras/wrappers.py:67]: Where are `optimizer` and `loss` used? Are they passed to the `compile` method?

File: scikeras/wrappers.py (hunk 7 above)
Comment (adriangb) [reply scikeras/wrappers.py:69]: No, we do not. Maybe I'll just remove that parameter for now, I don't want to go down that rabbit hole in this PR...

File: scikeras/wrappers.py (hunk 8 above)
Comment (adriangb) [reply scikeras/wrappers.py:67]: As it stands right now, it would be up to the user to declare their `build_fn` to use them. I.e.:

```python3
def build_fn(optimizer=..., loss=...):
    model = Model()
    ...
    model.compile(loss=loss, optimizer=optimizer)
```

**Comment by adriangb** [comment]
> There might be some missing parameters

The reasons they are not being included are:
* `steps_per_epoch`: the docs say `This argument is not supported with array inputs` so I think it only makes sense for `tf.data`, which we do not support here.
* `sample_weight`: this is/will still be handled in `fit` like Scikit-Learn does.
* `class_weight`: I think this one requires a bit of massaging to make Scikit-Learn's interpretation of this ([ref](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)) match up with Keras' ([ref](https://www.tensorflow.org/api_docs/python/tf/keras/Model)). I think Keras maps "integer class indices" to weights while Scikit-Learn maps "class labels" to weights. I'd like to leave this for a separate PR.

Regarding `max_epochs`/`epochs`.
I think what Keras does is that it runs for `epochs-initial_epoch` epochs. Are you proposing that instead we just have a single parameter `max_epochs` and always start at epoch 0?
And separately, you are suggesting that `partial_fit` override whatever `max_epochs` was set to and pass `epochs=1` to `model_.fit`?

**Comment by stsievert** [comment]
> And separately, you are suggesting that partial_fit override whatever max_epochs was set to and pass epochs=1 to model_.fit?

Yes. Scikit-Learn's SGDClassifier/SGDRegressor/MLPClassifier do the same thing. Here's the docstring for [MLPClassifier.partial_fit][1]: "Update the model with a single iteration over the given data."

> I think what Keras does is that it runs for `epochs - initial_epoch` epochs. Are you proposing that instead we just have a single parameter max_epochs and always start at epoch 0?

I think the difference is semantic. I don't think it's super important. `max_epochs` provides a little more leeway to perform some early stopping. Sticking with the Keras API isn't a bad choice I think.

I mention it because Scikit-Learn calls this parameter `max_iter`. In the documentation they always mention `max_iter` is "the maximum number of passes over the training data (aka epochs)." 

[1]:https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html#sklearn.neural_network.MLPClassifier.partial_fit

**Comment by adriangb** [comment]
Hmm okay. I think what I'd like to do is keep this PR smallish and leave the following parameters for a future PR (that can also contain any logic related to their handling, like what you mention about `fit` and epochs).
* `class_weight`
* `epochs`/`initial_epoch`/`max_epochs`
So I'll remove `epochs` from here.
